File: Version0.1/Python/Gencode/genlib.py
import os
from sys import platform

def genlib(cpucompiler, gpucompiler, coredir):

    mydir = os.getcwd();
    os.chdir(coredir);

    if len(cpucompiler)>0:
        str = cpucompiler + " -fPIC -O3 -c commonCore.cpp";
        os.system(str);
        str = "ar rvs commonCore.a commonCore.o";
        os.system(str);

        str = cpucompiler + " -fPIC -O3 -c opuCore.cpp";
        os.system(str);
        str = "ar rvs opuCore.a opuCore.o";
        os.system(str);

    if len(gpucompiler)>0:
        str = gpucompiler + " -D_FORCE_INLINES -O3 -c --compiler-options '-fPIC' gpuCore.cu";
        os.system(str);
        str = "ar rvs gpuCore.a gpuCore.o";
        os.system(str);

    if platform == "darwin":
    #    mv("*.a",pwd() * "/Mac");
        if os.path.isdir("Mac")==False:
            os.mkdir("Mac")

        if os.path.isfile("commonCore.a"):
            os.system("mv commonCore.a Mac");

        if os.path.isfile("opuCore.a"):
            os.system("mv opuCore.a Mac");

        if os.path.isfile("cpuCore.a"):
            os.system("mv cpuCore.a Mac");

        if os.path.isfile("gpuCore.a"):
            os.system("mv gpuCore.a Mac");
    elif platform == "linux" or platform == "linux2":
        if os.path.isdir("Linux")==False:
            os.mkdir("Linux")

        if os.path.isfile("commonCore.a"):
            os.system("mv commonCore.a Linux");

        if os.path.isfile("opuCore.a"):
            os.system("mv opuCore.a Linux");

        if os.path.isfile("cpuCore.a"):
            os.system("mv cpuCore.a Linux");

        if os.path.isfile("gpuCore.a"):
            os.system("mv gpuCore.a Linux");

    elif platform == "win32":
        if os.path.isdir("Windows")==False:
            os.mkdir("Windows")

        if os.path.isfile("commonCore.a"):
            os.system("mv commonCore.a Windows");

        if os.path.isfile("opuCore.a"):
            os.system("mv opuCore.a Windows");

        if os.path.isfile("cpuCore.a"):
            os.system("mv cpuCore.a Windows");

        if os.path.isfile("gpuCore.a"):
            os.system("mv gpuCore.a Windows");

    # os.remove("commonCore.a");
    # os.remove("opuCore.a");
    # os.remove("cpuCore.a");
    # os.remove("gpuCore.a");
    # os.remove("commonCore.o");
    # os.remove("opuCore.o");
    # os.remove("cpuCore.o");
    # os.remove("gpuCore.o");

    os.chdir(mydir);

    return 0

File: lib/genlib.jl
function genlib(cpucompiler::String, gpucompiler::String)
# Examples: genlib("g++","");
#           genlib("g++","nvcc");

if length(cpucompiler)==0
    error("cpucompiler is empty");
end

str = `$cpucompiler -fPIC -O3 -c commonCore.cpp`;
run(str);
str = `ar rvs commonCore.a commonCore.o`;
run(str);

str = `$cpucompiler -fPIC -O3 -c opuCore.cpp`;
run(str);
str = `ar rvs opuCore.a opuCore.o`;
run(str);

# str = `$cpucompiler -fPIC -O3 -c cpuCore.cpp -fopenmp`;
# run(str);
# str = `ar rvs cpuCore.a cpuCore.o`;
# run(str);

if length(gpucompiler)>0
    str = `$gpucompiler -D_FORCE_INLINES -O3 -c --compiler-options '-fPIC' gpuCore.cu`;
    run(str);
    str = `ar rvs gpuCore.a gpuCore.o `;
    run(str);
end

if Sys.isapple()
#    mv("*.a",pwd() * "/Mac");
    if !isdir("Mac")
        mkdir("Mac")
    end
    if isfile("commonCore.a")
        run(`mv commonCore.a Mac`);
    end
    if isfile("opuCore.a")
        run(`mv opuCore.a Mac`);
    end
    if isfile("cpuCore.a")
        run(`mv cpuCore.a Mac`);
    end
    if isfile("gpuCore.a")
        run(`mv gpuCore.a Mac`);
    end
elseif Sys.isunix()
    if !isdir("Linux")
        mkdir("Linux")
    end
    if isfile("commonCore.a")
        run(`mv commonCore.a Linux`);
    end
    if isfile("opuCore.a")
        run(`mv opuCore.a Linux`);
    end
    if isfile("cpuCore.a")
        run(`mv cpuCore.a Linux`);
    end
    if isfile("gpuCore.a")
        run(`mv gpuCore.a Linux`);
    end
elseif Sys.iswindows()
    if !isdir("Windows")
        mkdir("Windows")
    end
    if isfile("commonCore.a")
        run(`mv commonCore.a Windows`);
    end
    if isfile("opuCore.a")
        run(`mv opuCore.a Windows`);
    end
    if isfile("cpuCore.a")
        run(`mv cpuCore.a Windows`);
    end
    if isfile("gpuCore.a")
        run(`mv gpuCore.a Windows`);
    end
end

end

# delete *.o;
# delete *.a;
